src/common/telemetry.ts

```diff
@@ -1,377 +1,381 @@
+        private static APPINSIGHTS_INSTRUMENTATIONKEY: string = "AIF-d9b70cd4-b9f9-4d70-929b-a071c400b217"; // Matches vscode telemetry key
+        private static REGISTRY_SQMCLIENT_NODE: string = "\\SOFTWARE\\Microsoft\\SQMClient";
+        private static REGISTRY_USERID_VALUE: string = "UserId";
+        private static INTERNAL_DOMAIN_SUFFIX: string = "microsoft.com";
+        private static INTERNAL_USER_ENV_VAR: string = "TACOINTERNAL";
+
+        private static get telemetrySettingsFile(): string {
+            let settingsHome = HostPlatform.getSettingsHome();
+            return path.join(settingsHome, TelemetryUtils.TELEMETRY_SETTINGS_FILENAME);
+        }
 
-            public static generateGuid(): string {
-                let hexValues: string[] = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F"];
-                // c.f. rfc4122 (UUID version 4 = xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx)
-                let oct: string = "";
-                let tmp: number;
-                /* tslint:disable:no-bitwise */
-                for (let a: number = 0; a < 4; a++) {
-                    tmp = (4294967296 * Math.random()) | 0;
-                    oct += hexValues[tmp & 0xF] + hexValues[tmp >> 4 & 0xF] + hexValues[tmp >> 8 & 0xF] + hexValues[tmp >> 12 & 0xF] + hexValues[tmp >> 16 & 0xF] + hexValues[tmp >> 20 & 0xF] + hexValues[tmp >> 24 & 0xF] + hexValues[tmp >> 28 & 0xF];
-                }
+        public static init(appVersion: string, initOptions: ITelemetryInitOptions): Q.Promise<any> {
+            TelemetryUtils.loadSettings();
 
-                // "Set the two most significant bits (bits 6 and 7) of the clock_seq_hi_and_reserved to zero and one, respectively"
-                let clockSequenceHi: string = hexValues[8 + (Math.random() * 4) | 0];
-                return oct.substr(0, 8) + "-" + oct.substr(9, 4) + "-4" + oct.substr(13, 3) + "-" + clockSequenceHi + oct.substr(16, 3) + "-" + oct.substr(19, 12);
-                /* tslint:enable:no-bitwise */
+            if (initOptions.isExtensionProcess) {
+                let TelemetryReporter = require("vscode-extension-telemetry").default;
+                Telemetry.reporter = new TelemetryReporter(Telemetry.appName, appVersion, TelemetryUtils.APPINSIGHTS_INSTRUMENTATIONKEY);
+            } else {
+                Telemetry.reporter = new ExtensionTelemetryReporter(Telemetry.appName, appVersion, TelemetryUtils.APPINSIGHTS_INSTRUMENTATIONKEY);
             }
 
-            public static getTelemetryOptInSetting(): boolean {
-                if (TelemetryUtils.telemetrySettings.optIn === undefined) {
-                    // Opt-in by default
-                    TelemetryUtils.telemetrySettings.optIn = true;
-                }
+            return Q.all([TelemetryUtils.getUserId()])
+            .spread<any>(function(userId: string): void {
+                TelemetryUtils.userId = userId;

```

Comment:
Are we intentionally generating / tracking user IDs when vscode does not? It looks like vscode tracks machine IDs and session IDs and sqm IDs, but not user IDs. Since we can't correlate this data with anything other than more of our data, it might not be worth including it.